hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala

```diff
@@ -218,19 +224,28 @@ class JSBuilder(using TL, State, Ctx) extends CodeBuilder:
                 .mkDocument(doc"")
             val preCtorCode = ctorAuxParams.flatMap(ps => ps).foldLeft(body(preCtor, endSemi = true)):
               case (acc, (sym, nme)) =>
-                doc"$acc # this.${sym.name} = $nme;"
-            val ctorCode = doc"$preCtorCode${body(ctor, endSemi = false)}"
+                doc"$acc # this${fieldSelect(sym.name)} = $nme;"
+            val ctorCode = doc"$preCtorCode${body(ctor, endSemi = auxParams.nonEmpty)}"
 
-            val ctorBod = if auxParams.isEmpty then
-              doc"${braced(ctorCode)}"
+            val ctorAux = if auxParams.isEmpty then
+              ctorCode
             else
               val pss = ctorAuxParams.map(_.map(_._2))
               val newCtorCode = doc"$ctorCode # return this;"
               val ctorBraced = doc"${ braced(newCtorCode) }"
               val funBod = pss.foldRight(ctorBraced):
                 case (psDoc, doc) => doc"(${psDoc.mkDocument(", ")}) => $doc"
 
-              doc"${ braced(doc" # return $funBod") }"
+              doc" # return $funBod"
+            
+            val ctorBod = if isModule then
+              ownr match
+              case S(owner) =>
+                braced(doc" # ${mkThis(owner)}.${sym.nme} = ${getVar(isym)};$ctorCode")
```

Comment:
This is a bit messy. We get code like:
```fs
//│     this.Baz = class Baz1 {
//│       static {
//│         this.Baz = Baz1;
```
where the outer assignment is completely redundant. We should only do the outer assignment if the inner one is not already performed.

Reply:
And actually, we have exactly the same redundancy in the other case:
```fs
//│ let Bar3;
//│ Bar3 = class Bar2 {
//│   static {
//│     Bar3 = Bar2;
```

Reply:
It caused problem if the class is at the top level. Seems that you're not allowed to have the same class name if it is just
```
class Baz { ... }
class Baz { ... }
```
whereas
```
Baz1 = class Baz { ... }
```
does not actually bind the name Baz

Reply:
Couldn't you just make it
```js
(class Baz { ... })
(class Baz { ... })
```

Reply:
right, it works